common/fs/src/cache/event_debouncer.rs

```diff
@@ -0,0 +1,407 @@
+
+#[derive(Default)]
+struct PrioritisedEventQueue<T: std::cmp::Eq + std::hash::Hash> {
+    pq: priority_queue::PriorityQueue<T, Reverse<i64>>,
+}
+
+impl<T: std::cmp::Eq + std::hash::Hash> PrioritisedEventQueue<T> {
+    fn new() -> Self {
+        Self {
+            pq: priority_queue::PriorityQueue::default(),
+        }
+    }
+
+    #[cfg(test)]
+    fn len(&self) -> usize {
+        self.pq.len()
+    }
+}
+
+impl PrioritisedEventQueue<PrioritisedEvents<(WatchEvent, i64)>> {
+    // Event should go on the back of the queue
+    fn insert_event(&mut self, event: WatchEvent, ts: i64) {
+        let p_event_priority = event
+            .path_id()
+            .and_then(|key| self.pq.get_priority(key).copied());
+
+        if let Some(priority) = p_event_priority {
+            if let Some((p_event, _)) = event.path_id().and_then(|key| {
+                self.pq
+                    .change_priority(key, std::cmp::max(Reverse(ts), priority));
+                self.pq.get_mut(key)
+            }) {
+                p_event.update_events_from_back((event, ts));
+            }
+        } else if let Some(path) = event.path_id() {
+            self.pq.push(
+                PrioritisedEvents::new_with_event(path.clone(), (event, ts)),
+                Reverse(ts),
+            );
+        };
+    }
+
+    // Event should go on the front of the queue
+    fn requeue_event(&mut self, event: WatchEvent, ts: i64) {
+        let p_event_priority = event
+            .path_id()
+            .and_then(|key| self.pq.get_priority(key).copied());
+
+        if let Some(priority) = p_event_priority {
+            if let Some((p_event, _)) = event.path_id().and_then(|key| {
+                self.pq
+                    .change_priority(key, std::cmp::max(Reverse(ts), priority));
+                self.pq.get_mut(key)
+            }) {
+                p_event.update_events_from_front((event, ts));
+            }
+        } else if let Some(path) = event.path_id() {
+            self.pq.push(
+                PrioritisedEvents::new_with_event(path.clone(), (event, ts)),
+                Reverse(ts),
+            );
+        };
+    }
+
+    #[cfg(test)]
+    fn get(&self, key: &PathBuf) -> Option<(&PrioritisedEvents<(WatchEvent, i64)>, i64)> {
+        self.pq.get(key).map(|(ret, priority)| (ret, priority.0))
+    }
+
+    fn pop(&mut self) -> Option<(WatchEvent, i64)> {
+        let maybe_prioritised_event = self.pq.peek_mut();
+
+        let (ret, new_priority) = maybe_prioritised_event?.0.pop()?;
+        if let Some(key) = ret.0.path_id() {
+            self.pq
+                .change_priority(key, Reverse(new_priority.unwrap_or(i64::MAX)));
+        }
+        Some(ret)
+    }
+}
+
+#[derive(Debug)]
+struct PrioritisedEvents<T> {
+    key: std::path::PathBuf, // What filepath is this talking about
+    // Coalescing FIFO queue of events. end is head, start is tail
+    pub events: smallvec::SmallVec<[T; 4]>, // Debounced events
+}
```

Comment:
does it have any delay to keep some backlog of events ready for coalescing?

Reply:
> full history, in case we get a long repeating stream of "Create, Write, Delete"" 

how is this history synchronized with actual file state/content on the disk?